File: backend/app/models/rss.py
from typing import Optional, List
from datetime import datetime
from sqlmodel import Field, SQLModel, Relationship

class RSSFeed(SQLModel, table=True):
    """RSS订阅源模型"""
    id: Optional[int] = Field(default=None, primary_key=True)
    name: str = Field(index=True)
    url: str = Field(unique=True)
    description: Optional[str] = None
    enabled: bool = Field(default=True)
    last_check: Optional[datetime] = None
    check_interval: int = Field(default=30)  # 检查间隔（分钟）
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    
    # 关联
    rules: List["RSSRule"] = Relationship(back_populates="rss_feed")

class RSSRule(SQLModel, table=True):
    """RSS过滤规则模型"""
    id: Optional[int] = Field(default=None, primary_key=True)
    name: str
    keyword: str  # 关键词或正则表达式
    is_regex: bool = Field(default=False)  # 是否是正则表达式
    include: bool = Field(default=True)  # True表示包含，False表示排除
    enabled: bool = Field(default=True)
    created_at: datetime = Field(default_factory=datetime.utcnow)
    
    # 外键
    rss_feed_id: int = Field(foreign_key="rssfeed.id")
    
    # 关联
    rss_feed: RSSFeed = Relationship(back_populates="rules")

class RSSEntry(SQLModel, table=True):
    """RSS条目记录模型，用于记录已处理的条目"""
    id: Optional[int] = Field(default=None, primary_key=True)
    entry_id: str = Field(index=True)  # 条目唯一标识（如guid）
    title: str
    link: str
    published: Optional[datetime] = None
    processed_at: datetime = Field(default_factory=datetime.utcnow)
    downloaded: bool = Field(default=False)
    
    # 外键
    rss_feed_id: int = Field(foreign_key="rssfeed.id")

# Pydantic模型，用于API
class RSSFeedCreate(SQLModel):
    """创建RSS订阅源的输入模型"""
    name: str
    url: str
    description: Optional[str] = None
    enabled: bool = True
    check_interval: int = 30

class RSSFeedUpdate(SQLModel):
    """更新RSS订阅源的输入模型"""
    name: Optional[str] = None
    url: Optional[str] = None
    description: Optional[str] = None
    enabled: Optional[bool] = None
    check_interval: Optional[int] = None

class RSSRuleCreate(SQLModel):
    """创建RSS规则的输入模型"""
    name: str
    keyword: str
    is_regex: bool = False
    include: bool = True
    enabled: bool = True
    
class RSSRuleUpdate(SQLModel):
    """更新RSS规则的输入模型"""
    name: Optional[str] = None
    keyword: Optional[str] = None
    is_regex: Optional[bool] = None
    include: Optional[bool] = None
    enabled: Optional[bool] = None 

File: backend/internal/model/rss.go
package model

import "time"

// RSSFeed RSS订阅源数据库模型
type RSSFeed struct {
	ID            uint      `gorm:"primaryKey" json:"id"`
	Name          string    `gorm:"index;not null" json:"name"`
	URL           string    `gorm:"uniqueIndex;not null" json:"url"`
	Description   *string   `json:"description"`
	Enabled       bool      `gorm:"default:true" json:"enabled"`
	LastCheck     *time.Time `json:"last_check"`
	CheckInterval int       `gorm:"default:30" json:"check_interval"`
	CreatedAt     time.Time `json:"created_at"`
	UpdatedAt     time.Time `json:"updated_at"`

	Aggregate bool   `gorm:"default:false" json:"aggregate"`
	Parser    string `gorm:"default:'mikan'" json:"parser"`

	Rules []RSSRule `gorm:"foreignKey:RSSFeedID;constraint:OnDelete:CASCADE" json:"rules,omitempty"`
}

func (RSSFeed) TableName() string { return "rssfeed" }

// RSSRule RSS过滤规则数据库模型
type RSSRule struct {
	ID        uint      `gorm:"primaryKey" json:"id"`
	Name      string    `gorm:"not null" json:"name"`
	Keyword   string    `gorm:"not null" json:"keyword"`
	IsRegex   bool      `gorm:"default:false" json:"is_regex"`
	Include   bool      `gorm:"default:true" json:"include"`
	Enabled   bool      `gorm:"default:true" json:"enabled"`
	CreatedAt time.Time `json:"created_at"`
	RSSFeedID uint      `gorm:"index;not null" json:"rss_feed_id"`
	RSSFeed   *RSSFeed  `gorm:"foreignKey:RSSFeedID" json:"-"`
}

func (RSSRule) TableName() string { return "rssrule" }

// RSSEntry RSS条目记录数据库模型
type RSSEntry struct {
	ID          uint       `gorm:"primaryKey" json:"id"`
	EntryID     string     `gorm:"index;not null" json:"entry_id"`
	Title       string     `gorm:"not null" json:"title"`
	Link        string     `gorm:"not null" json:"link"`
	Published   *time.Time `json:"published"`
	ProcessedAt time.Time  `json:"processed_at"`
	Downloaded  bool       `gorm:"default:false" json:"downloaded"`
	RSSFeedID   uint       `gorm:"index;not null" json:"rss_feed_id"`

	// 解析后字段（用于诊断/筛选）
	ParsedAnime   *string `json:"parsed_anime,omitempty"`
	ParsedEpisode *int    `json:"parsed_episode,omitempty"`
	ParsedGroup   *string `json:"parsed_group,omitempty"`
	MatchedAnimeID *uint  `gorm:"index" json:"matched_anime_id,omitempty"`
}

func (RSSEntry) TableName() string { return "rssentry" }

// RSSFeedCreate 创建RSS订阅源请求
type RSSFeedCreate struct {
	Name          string  `json:"name" binding:"required"`
	URL           string  `json:"url" binding:"required"`
	Description   *string `json:"description"`
	Enabled       *bool   `json:"enabled"`
	CheckInterval *int    `json:"check_interval"`
	Aggregate     *bool   `json:"aggregate"`
	Parser        *string `json:"parser"`
}

// RSSFeedUpdate 更新RSS订阅源请求
type RSSFeedUpdate struct {
	Name          *string `json:"name"`
	URL           *string `json:"url"`
	Description   *string `json:"description"`
	Enabled       *bool   `json:"enabled"`
	CheckInterval *int    `json:"check_interval"`
	Aggregate     *bool   `json:"aggregate"`
	Parser        *string `json:"parser"`
}

// RSSRuleCreate 创建RSS规则请求
type RSSRuleCreate struct {
	Name    string `json:"name" binding:"required"`
	Keyword string `json:"keyword" binding:"required"`
	IsRegex bool   `json:"is_regex"`
	Include bool   `json:"include"`
	Enabled bool   `json:"enabled"`
}

// RSSRuleUpdate 更新RSS规则请求
type RSSRuleUpdate struct {
	Name    *string `json:"name"`
	Keyword *string `json:"keyword"`
	IsRegex *bool   `json:"is_regex"`
	Include *bool   `json:"include"`
	Enabled *bool   `json:"enabled"`
}
